examples.py:

jeff922 opened the pull request:
vector_uxg_pdw_example() - Added third pdw to example.  Corrected stream_start to stream_play

  vector_uxg_lan_streaming_example() - added marker output to easily trigger on signal, added PDW pulses, moved RF on point later in code to avoid CW tone output before streaming.

  analog_uxg_pdw_example() - Add pdws with frequency phase coding table indexes.

instruments.py:
  AnalogUXG - _init_ - Added N5193A check and mode change to turn on streaming.  Avoids exception if streaming was not on.
  Added padding block method
  Added bin_freqPhaseCodingSingleEntry and bin_freqPhaseCodingBlock methods.
  bin_pdw_file_builder() - Added links to use FPC coding in example.  Includes flag to disable FPC if needed to revert to legacy behavior.  Corrected offset shift direction << to >> and updated behavior to handle fpc blocks > 4096 bytes.
  VectorUXG - stream_configure() - Corrected trigOut to trigOutPort
  VectorUXG - bin_pdw_builder3() - added _wfmType=0 to correct missing variable.

Commit:
examples.py:
  vector_uxg_pdw_example() - Added third pdw to example.  Corrected stream_start to stream_play

  vector_uxg_lan_streaming_example() - added marker output to easily trigger on signal, added PDW pulses, moved RF on point later in code to avoid CW tone output before streaming.

  analog_uxg_pdw_example() - Add pdws with frequency phase coding table indexes.

instruments.py:
  AnalogUXG - _init_ - Added N5193A check and mode change to turn on streaming.  Avoids exception if streaming was not on.
  Added padding block method
  Added bin_freqPhaseCodingSingleEntry and bin_freqPhaseCodingBlock methods.
  bin_pdw_file_builder() - Added links to use FPC coding in example.  Includes flag to disable FPC if needed to revert to legacy behavior.  Corrected offset shift direction << to >> and updated behavior to handle fpc blocks > 4096 bytes.
  VectorUXG - stream_configure() - Corrected trigOut to trigOutPort
  VectorUXG - bin_pdw_builder3() - added _wfmType=0 to correct missing variable.

morgan-at-keysight commented on pyarbtools/instruments.py:
```diff
@@ -1384,6 +1392,138 @@ def bin_pdw_builder(self, operation=0, freq=1e9, phase=0, startTimeSec=0, width=
+             TODO - Combination of simultaneous phase and frequency modulation not yet implemented
+        """
+        if ((len(hexPatternString) % 2) != 0):
+            raise Exception('Hex pattern length must be a multiple of 2: Length is ' + str(len(hexPatternString)))
+
+        hexPatternBytes = bytearray.fromhex(hexPatternString)
+        numBitsInPattern = 8 * len(hexPatternBytes)
+
+        if (codingType !=0 and codingType !=1):
+            raise Exception('Only phase and frequency coding via streaming has been implemented in this example')
+        if (numBitsPerSubpulse != 1):
+            raise Exception('Only one bit per subpulse has been implemented in this example')
+        if (len(hexPatternBytes) > 8192):
+            raise Exception('Pattern must be less than 8192 bytes')
+        if (len(comment) > 60):
+            raise Exception('Comment must be less than 60 characters long')
```
It's best practice to use a specific type of exception rather than the generic "Exception" class. I've created a few different types of exception classes that point to a specific problem. See errors.py

Suggestion: change Exception to error.UXGError.

jeff922 replied:
Agreed.  Should each line be changed from "raise Exception" to "raise error.UXGError"?  How would I correct my committed fork?  Re-commit and submit another pull request?

morgan-at-keysight replied:
Yes, that's the correct fix. Based on my reading, you should be able to make changes and then commit them and it will automatically update the pull request.

morgan-at-keysight commented on pyarbtools/instruments.py:
```diff
@@ -1401,36 +1541,46 @@ def bin_pdw_file_builder(self, pdwList):
             (bytes): Binary data that contains a full PDW file that can
                 be downloaded to and played out of the UXG.
         """
+        #Include frequency phase coding block flag: 1 = yes, 0 = no
+        includeFpcBlock = 1
```
Should we make includeFpcBlock an argument to this function rather than a hard-coded value inside the function?

jeff922 replied:
Possibly.  The fpcblock flag could be removed entirely.  I included it to emulate the older behavior if needed.  Options would be to remove the flag check and always include fpc block or add an argument to pass the flag.  Do you have a preference?

morgan-at-keysight replied:
I think this might go deeper than just handling the flag. I now see that setting the flag results in calling the function to build the FPC block. The modular nature of Pyarbtools suggests that the FPC building function should be _external_ to the PDW building function so the two can be defined independently.
For now, let's leave it the way it is since it doesn't break anything and doesn't expose a new function for documentation. We can revisit the structure after the next update.

morgan-at-keysight commented on pyarbtools/instruments.py:
```diff
@@ -1401,36 +1541,46 @@ def bin_pdw_file_builder(self, pdwList):
-        # 4016 bytes of padding ensures that the first PDw begins @ byte 4097
-        padData = (0).to_bytes(4016, byteorder='little')
-        padding = [padBlockId, res3, size, padData]
+        # FPC Block - skip fpcBlock if flag is zero
+        fpcBlock = [b'']
+        if (includeFpcBlock):
+            fpcBlock = self.bin_pdw_freqPhaseCodingBlock()
+        fpcBlockSize = len(b''.join(fpcBlock))
 
-        # PDW block
+        #PDW block header must start at byte 4080 so PDW stream data starts at byte 4097
+        paddingSize = 4080 - tempHeaderSize - fpcBlockSize
+        paddingBlock = self.paddingBlock(paddingSize)
+
+        # PDW block header = 16 bytes
         pdwBlockId = (16).to_bytes(4, byteorder='little')
         res4 = (0).to_bytes(4, byteorder='little')
         pdwSize = (0xffffffffffffffff).to_bytes(8, byteorder='little')
         pdwBlock = [pdwBlockId, res4, pdwSize]
 
         # Build PDW file from header, padBlock, pdwBlock, and PDWs
-        pdwFile = header + padding + pdwBlock
+        pdwFile = header + fpcBlock + paddingBlock + pdwBlock
+
```
This is an elegant solution to adding the FPC block into a PDW. Nice work.

jeff922 replied:
Thanks. We could also incorporate the padding block method into bin_pdw_file_builder in the VectorUXG class in instruments.py as well.  I didn't want to make changes without consulting you first.

morgan-at-keysight replied:
I think that's a good idea, but let's add it in once we've fully addressed the structure of defining the FPC and incorporating it into generic PDWs.

morgan-at-keysight commented on pyarbtools/examples.py:
```diff
@@ -273,10 +274,14 @@ def vector_uxg_pdw_example(ipAddress):
 
     uxg = pyarbtools.instruments.VectorUXG(ipAddress, port=5025, timeout=10, reset=True)
-    uxg.configure(rfState=1, modState=1)
+    uxg.configure(rfState=0, modState=1)
```
Nice catch.

morgan-at-keysight commented on pyarbtools/examples.py:
```diff
@@ -348,11 +362,15 @@ def analog_uxg_pdw_example(ipAddress):
-               [2, 1e9, 0, 20e-6, 1e-6, 1, 0, 2, 0, 0, 0, 0, 0, 0]]
+
+    #Last PDW with operation = 2 does not play since it is marked as end of PDW list
+    pdwList = [[1, 980e6, 0, 0    , 10e-6, 1, 0, 2, 0, 0, 3, 0, 4000000, 0],
+               [0, 1e9  , 0, 20e-6, 15e-6 , 1, 0, 2, 0, 0, 0, 1, 0,       0],
+               [0, 1.01e9, 0, 40e-6, 20e-6, 1, 0, 2, 0, 0, 0, 2, 0, 0],
+               [2, 1e9,   0, 80e-6, 5e-6, 1, 0, 2, 0, 0, 0, 1, 0, 0]]
```
Very minor, but spacing got changed so the fields don't line up anymore.

morgan-at-keysight commented on pyarbtools/examples.py:
```diff
@@ -256,15 +256,16 @@ def vector_uxg_pdw_example(ipAddress):
 
     uxg.csv_pdw_file_download(pdwName, fields, data)
 
-    uxg.stream_start(pdwID=pdwName)
+    uxg.stream_play(pdwID=pdwName)
```
Another nice catch.

morgan-at-keysight reviewed (changes requested):
This is a great addition to Pyarbtools, thank you for contributing! You've maintained consistent code styling, which is very helpful for readability (and sort of uncommon, so thank you).

I've made a couple comments and small change requests below, but once those are fixed I'll merge the code. 

One question about analog_uxg_pdw_example. I see that the bin_freqPhaseCodingBlock functions are called within the bin_pdw_builder functions, but it wasn't immediately clear that the FPC block was activated in the example function. Is there a way to define and include an FPC block in a PDW stream/file similar to the way you can define a PDW by sending a list of fields to the bin_pdw_file_builder function? This can be adjusted in a future commit.

Commit:
examples.py:
  vector_uxg_pdw_example() - Added third pdw to example.  Corrected stream_start to stream_play

  vector_uxg_lan_streaming_example() - added marker output to easily trigger on signal, added PDW pulses, moved RF on point later in code to avoid CW tone output before streaming.

  analog_uxg_pdw_example() - Add pdws with frequency phase coding table indexes.

instruments.py:
  AnalogUXG - _init_ - Added N5193A check and mode change to turn on streaming.  Avoids exception if streaming was not on.
  Added padding block method
  Added bin_freqPhaseCodingSingleEntry and bin_freqPhaseCodingBlock methods.
  bin_pdw_file_builder() - Added links to use FPC coding in example.  Includes flag to disable FPC if needed to revert to legacy behavior.  Corrected offset shift direction << to >> and updated behavior to handle fpc blocks > 4096 bytes.
  VectorUXG - stream_configure() - Corrected trigOut to trigOutPort
  VectorUXG - bin_pdw_builder3() - added _wfmType=0 to correct missing variable.

morgan-at-keysight reviewed (approved):
Nice work. Thanks for bearing with me as I figure out how to handle pull requests!
